EZP-27948: When swapping Locations with the same name and parent, URL aliases are not swapped as well

natanael89 opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-27948](https://jira.ez.no/browse/EZP-27948)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `6.x`/`7.x`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | probably yes (see below)

Description from JIRA:
> Currently, when swapping Locations with the same name and parent, URL aliases are not swapped as well. This results in situations when the Editor cannot effectively swap the Locations for a Content.
This behaviour was correct in eZ Publish 5.4.

The cause of the issue is that swapping Locations in conjunction with swapping UrlAliases is effectively cancelling each other (in this particular use-case; in others it works fine). To circumvent this, I made two changes:
1. Instead of using UrlAlias text from the database, I republish UrlAliases for the swapped Locations using the Contents' names. This allows me to get rid of the numbers at the end of UrlAliases (if they are no longer necessary).
2. I detect this particular use-case and make sure the UrlAliases are then published in the correct order. For other use-cases, the order is not important.

Some examples how this affects behaviour:
1. There are two Content Objects named "Test". The first is accessible via "/Test" and the second one via "/Test2". We are swapping their Locations.
* Previously nothing changed.  The first is still accessible via "/Test" and the second one via "/Test2".
* With this PR, the first is accessible via "/Test2" and the second one via "/Test".
2. There are two Content Objects. The first is named "Test" and is accessible via "/Test". The second is named "Another test" and is accessible via "/Parent/Another-test2". We are swapping their Locations.
* Previously the first would be accessible via "/Parent/Test". The second would be accessible via "/Another-test2".
* With this PR, the first is accessible via "/Parent/Test" (no change). The second is accessible via "/Another-test" (no "2" at the end, since it is no longer necessary in the new Location).
3. There are two Content Objects named "Test". The first is accessible via "/Test3" and the second one via "/Test4". There are no other Content Objects named "Test" under root (we can guess that there were, but they were moved or renamed). We are swapping Locations of these two Content Objects.
* Previously nothing changed.  The first is still accessible via "/Test3" and the second one via "/Test4".
* With this PR, the first is accessible via "/Test2" and the second one via "/Test". "/Test3" now redirects to "/Test" and "/Test4" redirects to "/Test2". So, old URLs are redirecting to the new URLs of swapped Content, which seems correct.
4. There are two Content Objects named "Test". The first is accessible via "/Test2" and the second one via "/Test3". There are no other Content Objects named "Test" under root (we can guess that there were, but they were moved or renamed). We are swapping Locations of these two Content Objects.
* Previously nothing changed.  The first is still accessible via "/Test2" and the second one via "/Test3".
* With this PR, the first is accessible via "/Test2" and the second one via "/Test". "/Test3" now redirects to "/Test2". However, URL "/Test2" is still referencing the first Content Object after the swap, which might be misleading and probably should be documented if the PR is merged.

An alternative approach to the one suggested in this PR would be to keep using UrlAlias text from the database and not generate a new one from the Content's name. Instead, detect the problematic use-case and then republish UrlAliases using their original UrlAlias text (not swapped). This, however, would still require Content Gateway inside UrlAlias Handler, so it would not make the solution much simpler. Also, I think that getting rid of the unnecessary numbers at the end of UrlAliases during the swap is something good because they will be cut off anyway on Content's republish.

Another alternative I can think of would be to not historize all UrlAliases before the swap but only historize selectively when needed. This however would require changing the way historizing works and will be definitely less efficient since it would require to historize step by step and only for one language at the time. UrlAliases are currently stored per language mask, so it would most probably require a lot of operations on the database or a more complicated way of calculating them.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

natanael89 commented on eZ/Publish/SPI/Persistence/Content/UrlAlias/Handler.php:
```diff
@@ -149,10 +149,19 @@ public function locationCopied($locationId, $newLocationId, $newParentId);
+        $location2Id,
+        $location2ParentId,
+        $content2Id
+    );
```
An alternative would be to add method `loadContentInfoByLocation()` or `loadContentByLocation()` to `eZ\Publish\Core\Persistence\Legacy\Content\Gateway` interface. What do you think about this solution?

alongosz replied:
> An alternative would be to add method loadContentInfoByLocation() or loadContentByLocation() to eZ\Publish\Core\Persistence\Legacy\Content\Gateway

Yep, we can mess with that one without worries. But first please see if some Gateway already implements something that could be reused (I didn't check).

natanael89 replied:
I created new method since there wasn't any to reuse.

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -628,46 +639,122 @@ public function locationSwapped($location1Id, $location1ParentId, $location2Id,
+     *
+     * @return array
+     */
+    private function getNamesByLanguage($contentInfo)
```
nitpick: sounds more like a `getNamesInAllLanguages()`

natanael89 replied:
Thanks, I renamed it to `getNamesForAllLanguages()`.

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -693,6 +780,68 @@ private function historizeBeforeSwap($location1Entries, $location2Entries)
+     * @param array $locationEntries
+     * @param int $languageId
+     *
+     * @return array
```
Nitpick: Return type should be `array|null`

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -693,6 +780,68 @@ private function historizeBeforeSwap($location1Entries, $location2Entries)
+     *
+     * @return array
+     */
+    private function getLocationEntryInLanguage($locationEntries, $languageId)
```
Nitpick: Missing `array` type hint for `$locationEntries` parameter

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -628,46 +639,122 @@ public function locationSwapped($location1Id, $location1ParentId, $location2Id,
+        $names1 = $this->getNamesForAllLanguages($contentInfo1);
+        $names2 = $this->getNamesForAllLanguages($contentInfo2);
+
+        $alwaysAvailable1 = (bool)($contentInfo1['language_mask'] & 1);
```
@alongosz Is there an better way to check this "always available" flag? I mean constant representing this flag or method  which might be used to check this.

alongosz replied:
>  Is there an better way to check this "always available" flag? I mean constant representing this flag or method which might be used to check this.

Hmm, here it was hardcoded from the beginning. I think that maybe we should start relying more on `\eZ\Publish\Core\Persistence\Legacy\Content\Language\MaskGenerator::isAlwaysAvailable`.

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php:
```diff
@@ -927,7 +928,23 @@ public function loadContentInfo($contentId)
+     */
+    public function loadContentInfoByLocationId($locationId)
+    {
+        return $this->internalLoadContentInfo('main_node_id', 'ezcontentobject_tree', $locationId);
```
and now going back to usage of arguments inside this method:
`$query->bindValue($id, null, $column === 'id' ? PDO::PARAM_INT : PDO::PARAM_STR)`
we have `main_node_id` which also should be PARAM_INT. Hence my previous comment :)

So let the caller decide on where clause, let's not try to combine it from atomic arguments.

Suggestion: maybe deciding what kind of argument this method needs would be simpler if we used QueryBuilder directly from Doctrine\DBAL, not the Zeta Database components.

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -693,6 +780,68 @@ private function historizeBeforeSwap($location1Entries, $location2Entries)
+     * @param string|int $location1ParentId
+     * @param string $name1
+     * @param array $location2Entries
+     * @param string|int $location2ParentId
```
Please use single type-hint. `int` in both cases.

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -628,46 +639,122 @@ public function locationSwapped($location1Id, $location1ParentId, $location2Id,
+            if (isset($names1[$languageCode], $names2[$languageCode])) {
+                $name1 = $names1[$languageCode];
+                $name2 = $names2[$languageCode];
+                if ($this->shouldUrlAliasForSecondLocationBePublishedFirst(
```
This whole loop contains too many code repetitions. What you need here is a method that returns an ordered list of sets of arguments to be passed to `internalPublishUrlAliasForLocation`. It could be some internal DTO ValueObject or just an associative array.

Maybe something like (adjust naming):
```php
$locationUrlAliasesList = $this->geUrlAliasesToBePublishedForLocations();
```
and internally instead of checking `shouldUrlAliasForSecondLocationBePublishedFirst` just "inline" that code and put proper data on the resulting list in a proper order. Don't forget about docblock describing what `shouldUrlAliasForSecondLocationBePublishedFirst` does, It's great :)

Please contact me if something here is not clear.

natanael89 commented:
@alongosz I made some changes according to your suggestions. Thanks :)
I left out the last one (about code repetitions in the loop) for the reasons we have discussed on slack (there is no good way to avoid repetitions completely, we could just move them elsewhere, but it would also probably increase the complexity).

alongosz commented:
Two improvements from my side:
- 5b068ee simplifies checking if a given language Id exists in a mask
- 66e025d solution which reduces number of lines of code to maintain ;)

I'm introducing DTOs inside single layer (added 2 new classes, marked as `@internal`) to improve readability. They're about transferring complex information between subsequent calls instead of operating on multiple variables in a single big code block.
Since we don't usually do it inside one layer, I'll understand if you don't like it :)

Reducing a number of lines of code could be done of course w/o these DTOs, it's all about readability. At least for me ;)

ping @natanael89 @adamwojs @andrerom

natanael89 commented:
@alongosz Looks good to me :)

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php:
```diff
@@ -912,7 +896,22 @@ private function internalLoadContentInfo($column, $tableName, $id)
     public function loadContentInfo($contentId)
     {
-        return $this->internalLoadContentInfo('id', 'ezcontentobject', $contentId);
+        $columnName = 'id';
```
It's more readable if you omit the variable now.

andrerom replied:
That said, why was this changed to inject query? _(not against, just curious)_

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/DTO/SwappedLocationProperties.php:
```diff
@@ -0,0 +1,56 @@
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\Core\Persistence\Legacy\Content\UrlAlias\DTO;
```
@alongosz Should we expand DTO to `DataTransfer` or similar?

alongosz replied:
> @alongosz Should we expand DTO to DataTransfer or similar?

@andrerom If you really want... For me DTO acronym is as clear as API or SPI.

lserwatka commented:
@alongosz @adamwojs @Nattfarinn we have to move forward with this PR. Please have a look.

lserwatka commented:
@micszo we will need an extensive QA effort on this PR.

lserwatka commented:
@natanael89 Could you clean commits here?

natanael89 commented:
@lserwatka I squashed my commits and rebased the branch to the latest 6.7. I left @alongosz's commits separate for clarity, but I can also squash them if needed.

lserwatka commented:
@natanael89 please merge this up to master with 6.13, 7.1

natanael89 commented:
Merged up to 6.13, 7.1 and master.

alongosz commented:
Thank you @natanael89. Great work :tada:
